sql: improve setseed builtin

yuzefovich opened the pull request:
This commit improves https://github.com/cockroachdb/cockroach/commit/b51da9ebe2d1c2e4b7ecf810edb5c88cdc991adb in
a couple of ways.

First, it makes the allocation of the RNG lazy, whenever either random()
or setseed() is called for the first time.

Here is the impact on `flowinfra/BenchmarkInfrastructure`:
```
name                         old time/op    new time/op    delta
Infrastructure/n1/r1-24        37.0µs ± 1%    23.0µs ± 2%  -37.74%  (p=0.000 n=10+10)
Infrastructure/n1/r100-24      87.7µs ± 1%    75.7µs ± 2%  -13.76%  (p=0.000 n=8+10)
Infrastructure/n1/r10000-24    5.03ms ± 2%    5.06ms ± 1%     ~     (p=0.165 n=10+10)
Infrastructure/n3/r1-24         749µs ± 4%     714µs ± 4%   -4.57%  (p=0.000 n=10+10)
Infrastructure/n3/r100-24      1.01ms ± 3%    0.96ms ± 4%   -5.09%  (p=0.000 n=10+10)
Infrastructure/n3/r10000-24    19.4ms ± 1%    19.6ms ± 1%     ~     (p=0.247 n=10+10)

name                         old alloc/op   new alloc/op   delta
Infrastructure/n1/r1-24        28.2kB ± 2%    22.3kB ± 1%  -20.77%  (p=0.000 n=9+8)
Infrastructure/n1/r100-24      74.2kB ± 0%    68.4kB ± 0%   -7.81%  (p=0.000 n=8+8)
Infrastructure/n1/r10000-24    4.58MB ± 0%    4.57MB ± 0%     ~     (p=0.161 n=8+8)
Infrastructure/n3/r1-24         168kB ± 1%     153kB ± 9%   -8.85%  (p=0.000 n=8+9)
Infrastructure/n3/r100-24       401kB ± 0%     383kB ± 1%   -4.47%  (p=0.000 n=8+9)
Infrastructure/n3/r10000-24    21.9MB ± 0%    22.0MB ± 2%     ~     (p=0.481 n=8+9)

name                         old allocs/op  new allocs/op  delta
Infrastructure/n1/r1-24           122 ± 3%       116 ± 1%   -4.64%  (p=0.000 n=9+8)
Infrastructure/n1/r100-24         236 ± 0%       230 ± 0%   -2.28%  (p=0.000 n=8+8)
Infrastructure/n1/r10000-24     10.4k ± 0%     10.4k ± 0%     ~     (p=0.151 n=8+8)
Infrastructure/n3/r1-24         1.52k ± 1%     1.50k ± 2%   -1.47%  (p=0.005 n=8+8)
Infrastructure/n3/r100-24       2.20k ± 0%     2.17k ± 1%   -1.16%  (p=0.000 n=8+9)
Infrastructure/n3/r10000-24     54.0k ± 0%     54.6k ± 5%     ~     (p=0.743 n=8+9)
```

Second, it fixes the logic so that all consequent statements on the same
session do use the right seed - previously, subqueries would use
a separate RNG created in `initEvalCtx`.

This refactor also allows us to only set the RNG field on the eval
context on the main code path.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/121523)
<!-- Reviewable:end -->

yuzefovich commented:
I think the only concern with this change is whether we need to make access to the RNG atomic. `rand.Rand` object itself is not concurrency safe IIRC, so before this change calls on `evalCtx.RNG` already could be racy; this change makes also the instantiation of `.rng` racy.

rafiss reviewed (changes requested):
nice find! i didn't realize that could be expensive.

i left some comments, and had one other thought:

another similar field in evalCtx is `ULIDEntropy`, which is initialized as `ulid.Monotonic(crypto_rand.Reader, 0)`. that constructor does not look expensive, but it might have the same issue with subqueries. actually, it's possible using `crypto_rand.Reader` there is also a bug, since it probably should be controlled by the same RNG that setseed controls.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-:-NuS3ocTDiE7Subarzos:bq40de1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/conn_executor.go` line 3711 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuS-SbN8miB0haFQwzL:-NuS-SbN8miB0haFQwzM:brjhb09) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e9baeda28603b8e2f795203a01efd22f1ee76c/pkg/sql/conn_executor.go#L3711)):*
> ```Go
> 
> func (ex *connExecutor) getSessionRNG() *rand.Rand {
> 	if ex.rng.external == nil {
> ```

we could use `sync.Once` to avoid the raciness, while still being cheap. it could be another field of the `rng` struct
___
*[`pkg/sql/schemachanger/scbuild/tree_context_builder.go` line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuS0u9f61D9Xw8NHU0T:-NuS0u9f61D9Xw8NHU0U:b-ntc3m0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e9baeda28603b8e2f795203a01efd22f1ee76c/pkg/sql/schemachanger/scbuild/tree_context_builder.go#L69)):*
> ```Go
> 		DescIDGenerator:      d.DescIDGenerator(),
> 		ULIDEntropy:          ulid.Monotonic(crypto_rand.Reader, 0),
> 		GetSessionRNG: func() *rand.Rand {
> ```

nit: instead of all of these having the same inlined function, we could make a reusable struct in some common package. (and we also may as well make it use sync.Once as well, if raciness is a concern).

```
type RNGFactory struct {
	once sync.Once
	rng  *rand.Rand
}

func NewRNGFactory() *RNGFactory {
	return &RNGFactory{}
}

func (r *RNGFactory) GetOrCreate() *rand.Rand {
	if r.rng == nil {
		r.once.Do(func() {
			r.rng, _ = randutil.NewPseudoRand()
		})
	}
	return r.rng
}
```
___
*[`pkg/sql/sem/builtins/builtins.go` line 2111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuS1hs3CFJC_0vYU_ap:-NuS1hs3CFJC_0vYU_aq:b-655u98) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e9baeda28603b8e2f795203a01efd22f1ee76c/pkg/sql/sem/builtins/builtins.go#L2111)):*
> ```Go
> 					return nil, pgerror.Newf(pgcode.InvalidParameterValue, "setseed parameter %f is out of allowed range [-1,1]", seed)
> 				}
> 				evalCtx.SetSessionRNGSeed(int64(math.Float64bits(float64(seed))))
> ```

this appears to be the only usage of `SetSessionRNGSeed`

why can't we just do:
```
evalCtx.GetSessionRNG().Seed(int64(math.Float64bits(float64(seed))))
```


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the suggestions! I used them as a base for the next iteration - let me know what you think.

> i didn't realize that could be expensive.

It's not that expensive, but given that we sometimes are asked to optimize queries in ms range even 10us can add up :) Also note that this is a very targeted micro-benchmark which effectively runs all cases through the "remote flow" codepath (so it might not be a good representation of reality).

> another similar field in evalCtx is `ULIDEntropy` ...

Good point. I did a quick check on the impact of removing initialization of that field (on top of this PR):
```
name                         old time/op    new time/op    delta
Infrastructure/n1/r1-24        16.7µs ± 2%    15.5µs ± 1%   -7.09%  (p=0.000 n=10+10)
Infrastructure/n1/r100-24      56.2µs ± 2%    55.0µs ± 1%   -2.10%  (p=0.000 n=10+10)
Infrastructure/n1/r10000-24    3.76ms ± 1%    3.78ms ± 2%     ~     (p=0.133 n=9+10)
Infrastructure/n3/r1-24         525µs ± 5%     509µs ± 4%   -3.04%  (p=0.029 n=10+10)
Infrastructure/n3/r100-24       724µs ± 4%     716µs ± 4%     ~     (p=0.182 n=10+9)
Infrastructure/n3/r10000-24    14.9ms ± 2%    15.0ms ± 2%     ~     (p=0.089 n=10+10)

name                         old alloc/op   new alloc/op   delta
Infrastructure/n1/r1-24        22.3kB ± 0%    18.0kB ± 0%  -19.30%  (p=0.000 n=8+8)
Infrastructure/n1/r100-24      68.2kB ± 0%    64.0kB ± 1%   -6.18%  (p=0.000 n=8+9)
Infrastructure/n1/r10000-24    4.55MB ± 0%    4.55MB ± 0%     ~     (p=0.515 n=10+8)
Infrastructure/n3/r1-24         149kB ± 1%     136kB ± 1%   -8.96%  (p=0.000 n=8+8)
Infrastructure/n3/r100-24       381kB ± 1%     373kB ± 7%   -2.00%  (p=0.046 n=8+9)
Infrastructure/n3/r10000-24    21.8MB ± 0%    21.8MB ± 0%     ~     (p=0.931 n=9+9)

name                         old allocs/op  new allocs/op  delta
Infrastructure/n1/r1-24           116 ± 0%       113 ± 0%   -2.59%  (p=0.000 n=8+8)
Infrastructure/n1/r100-24         230 ± 0%       227 ± 0%   -1.30%  (p=0.000 n=8+8)
Infrastructure/n1/r10000-24     10.3k ± 0%     10.3k ± 0%     ~     (p=0.501 n=10+8)
Infrastructure/n3/r1-24         1.49k ± 0%     1.48k ± 0%   -0.67%  (p=0.000 n=8+8)
Infrastructure/n3/r100-24       2.16k ± 0%     2.16k ± 3%   -0.10%  (p=0.019 n=8+8)
Infrastructure/n3/r10000-24     53.6k ± 0%     53.6k ± 0%     ~     (p=0.666 n=9+9)
```
so it seems worth it to look into - for now I left a TODO.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-:-NuVN2qP0BWXERbuur1w:b-yhq4sc)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/conn_executor.go` line 3711 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuS-SbN8miB0haFQwzL:-NuVK5SU63njTaiY0zNT:bae2r7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e9baeda28603b8e2f795203a01efd22f1ee76c/pkg/sql/conn_executor.go#L3711)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

we could use `sync.Once` to avoid the raciness, while still being cheap. it could be another field of the `rng` struct
</blockquote></details>

I refactored the code to have a mutex in the eval context, so now both initialization of the RNG and its usage is concurrency safe. Only the initialization of the mutex itself is racy, which seems acceptable.
___
*[`pkg/sql/schemachanger/scbuild/tree_context_builder.go` line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuS0u9f61D9Xw8NHU0T:-NuVJtbv6gZi2ykaKnma:b-ytkoq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e9baeda28603b8e2f795203a01efd22f1ee76c/pkg/sql/schemachanger/scbuild/tree_context_builder.go#L69)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: instead of all of these having the same inlined function, we could make a reusable struct in some common package. (and we also may as well make it use sync.Once as well, if raciness is a concern).

```
type RNGFactory struct {
	once sync.Once
	rng  *rand.Rand
}

func NewRNGFactory() *RNGFactory {
	return &RNGFactory{}
}

func (r *RNGFactory) GetOrCreate() *rand.Rand {
	if r.rng == nil {
		r.once.Do(func() {
			r.rng, _ = randutil.NewPseudoRand()
		})
	}
	return r.rng
}
```
</blockquote></details>

Thanks, I used this suggestion as a base for the next iteration.
___
*[`pkg/sql/sem/builtins/builtins.go` line 2111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuS1hs3CFJC_0vYU_ap:-NuVJs23ArOYWDZKIRh4:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3e9baeda28603b8e2f795203a01efd22f1ee76c/pkg/sql/sem/builtins/builtins.go#L2111)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this appears to be the only usage of `SetSessionRNGSeed`

why can't we just do:
```
evalCtx.GetSessionRNG().Seed(int64(math.Float64bits(float64(seed))))
```
</blockquote></details>

Good point, done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
i feel like i might be missing an insight into how the code is organized, but some of this feels a bit more confusing to me now. i do like that the initialization of evalCtx is simpler now though.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-:-NuVes3B3mmJO1NYwVhp:b7fydd7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/conn_executor.go` line 1690 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuVcrXr0ch9Cnuhf8PP:-NuVcrXs3MEPQGkzaesB:b7isrgn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c39bacd595b054f529997945ff20349ee9096397/pkg/sql/conn_executor.go#L1690)):*
> ```Go
> 		internal *rand.Rand
> 		// external is used to power random() builtin.
> 		external eval.RNGFactory
> ```

it seems important that this is a value, and not a pointer; maybe that deserves a comment.
___
*[`pkg/sql/sem/eval/context.go` line 308 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuVdl9Q41lCDRtVK7yt:-NuVdl9Q41lCDRtVK7yu:bj0a8sa) ([raw file](https://github.com/cockroachdb/cockroach/blob/c39bacd595b054f529997945ff20349ee9096397/pkg/sql/sem/eval/context.go#L308)):*
> ```Go
> 		// internal is used if Factory is not set. This allows us to only
> 		// initialize Factory on the main code path.
> 		internal *rand.Rand
> ```

nit: while reading this i found it super confusing to distinguish this internal `*rand.Rand` from the  one in `connExecutor.rng.internal`. AFAICT, they actually are completely unrelated, right?
___
*[`pkg/sql/sem/eval/context.go` line 314 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuVc0h-4ToBJ9l14r5a:-NuVc0h0EIjX9D-C_Lr7:b-2cs90p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c39bacd595b054f529997945ff20349ee9096397/pkg/sql/sem/eval/context.go#L314)):*
> ```Go
> // RNGFactory is a utility struct that allows for lazy allocation of the RNG.
> type RNGFactory struct {
> 	rng *rand.Rand
> ```

i find this a bit harder to follow now. in this version `RNGFactory` is just a wrapper around `*rand.Rand`, and there's no helpers to initialize the rng. i might be missing something -- what does the RNGFactory buy us now?
___
*[`pkg/sql/sem/eval/context.go` line 323 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuVUjZ3EwOtCk4kx15E:-NuVUjZ3EwOtCk4kx15F:b-at1w3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c39bacd595b054f529997945ff20349ee9096397/pkg/sql/sem/eval/context.go#L323)):*
> ```Go
> 	// the Context by value which we do in many places.
> 	if ec.RNG.mu == nil {
> 		ec.RNG.mu = &syncutil.Mutex{}
> ```

hm not sure how much to be worried about this, but if the initialization of the mutex itself is racy, does that introduce the possibility of a deadlock? (e.g. if the reference changes in between the call to lock/unlock)
___
*[`pkg/sql/sem/eval/context.go` line 344 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuVUbDB07ggGfRVLVfN:-NuVUbDB07ggGfRVLVfO:b-byrrq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c39bacd595b054f529997945ff20349ee9096397/pkg/sql/sem/eval/context.go#L344)):*
> ```Go
> func (ec *Context) RandFloat64() float64 {
> 	ec.ensureRNGMu()
> 	ec.RNG.mu.Lock()
> ```

i'm more worried about this change now. it's using a mutex on a potentially hot path of query execution. i'm guessing a benchmark of `SELECT random() FROM generate_series(1, 1000000) ` would perform significantly worse now. a similar realistic example is a query that has `ORDER BY random()`

even if true, i'm not sure what to make of that though... it definitely is the case that `rand.Rand` is not supposed to be used concurrently.

relatedly, using the mutex here doesn't actually make things deterministic, right? (since there's no guarantees about the order it gets accessed by different subqueries.) 
___
*[`pkg/sql/sem/eval/context.go` line 349 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuVekn05bL2ZlHcYINN:-NuVekn05bL2ZlHcYINO:bnxvoc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c39bacd595b054f529997945ff20349ee9096397/pkg/sql/sem/eval/context.go#L349)):*
> ```Go
> }
> 
> // RandSeed updates the session's RNG to the provided seed.
> ```

i find this a bit confusing, since sometimes it actually is updating the seed of the `internal` RNG.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: improve setseed builtin

This commit improves b51da9ebe2d1c2e4b7ecf810edb5c88cdc991adb in
a couple of ways.

First, it makes the allocation of the RNG lazy, whenever either random()
or setseed() is called for the first time.

Here is the impact on `flowinfra/BenchmarkInfrastructure`:
```
name                         old time/op    new time/op    delta
Infrastructure/n1/r1-24        37.0µs ± 1%    23.0µs ± 2%  -37.74%  (p=0.000 n=10+10)
Infrastructure/n1/r100-24      87.7µs ± 1%    75.7µs ± 2%  -13.76%  (p=0.000 n=8+10)
Infrastructure/n1/r10000-24    5.03ms ± 2%    5.06ms ± 1%     ~     (p=0.165 n=10+10)
Infrastructure/n3/r1-24         749µs ± 4%     714µs ± 4%   -4.57%  (p=0.000 n=10+10)
Infrastructure/n3/r100-24      1.01ms ± 3%    0.96ms ± 4%   -5.09%  (p=0.000 n=10+10)
Infrastructure/n3/r10000-24    19.4ms ± 1%    19.6ms ± 1%     ~     (p=0.247 n=10+10)

name                         old alloc/op   new alloc/op   delta
Infrastructure/n1/r1-24        28.2kB ± 2%    22.3kB ± 1%  -20.77%  (p=0.000 n=9+8)
Infrastructure/n1/r100-24      74.2kB ± 0%    68.4kB ± 0%   -7.81%  (p=0.000 n=8+8)
Infrastructure/n1/r10000-24    4.58MB ± 0%    4.57MB ± 0%     ~     (p=0.161 n=8+8)
Infrastructure/n3/r1-24         168kB ± 1%     153kB ± 9%   -8.85%  (p=0.000 n=8+9)
Infrastructure/n3/r100-24       401kB ± 0%     383kB ± 1%   -4.47%  (p=0.000 n=8+9)
Infrastructure/n3/r10000-24    21.9MB ± 0%    22.0MB ± 2%     ~     (p=0.481 n=8+9)

name                         old allocs/op  new allocs/op  delta
Infrastructure/n1/r1-24           122 ± 3%       116 ± 1%   -4.64%  (p=0.000 n=9+8)
Infrastructure/n1/r100-24         236 ± 0%       230 ± 0%   -2.28%  (p=0.000 n=8+8)
Infrastructure/n1/r10000-24     10.4k ± 0%     10.4k ± 0%     ~     (p=0.151 n=8+8)
Infrastructure/n3/r1-24         1.52k ± 1%     1.50k ± 2%   -1.47%  (p=0.005 n=8+8)
Infrastructure/n3/r100-24       2.20k ± 0%     2.17k ± 1%   -1.16%  (p=0.000 n=8+9)
Infrastructure/n3/r10000-24     54.0k ± 0%     54.6k ± 5%     ~     (p=0.743 n=8+9)
```

Second, it fixes the logic so that all consequent statements on the same
session do use the right seed - previously, subqueries would use
a separate RNG created in `initEvalCtx`.

This refactor also allows us to only set the RNG field on the eval
context on the main code path.

Release note: None

yuzefovich reviewed (commented):
Ok, I removed all new synchronization and made the change closer to what you originally suggested. Should be simpler now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-:-Nu_iHVV1DzIypEVx_Xn:bdwv5jj)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/conn_executor.go` line 1690 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuVcrXr0ch9Cnuhf8PP:-Nu_e-Hz3WjrluQZ6kwg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c39bacd595b054f529997945ff20349ee9096397/pkg/sql/conn_executor.go#L1690)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it seems important that this is a value, and not a pointer; maybe that deserves a comment.
</blockquote></details>

Done.
___
*[`pkg/sql/sem/eval/context.go` line 308 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuVdl9Q41lCDRtVK7yt:-Nu_e3Cq6vU64LyZc5Pj:b-ij2jp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c39bacd595b054f529997945ff20349ee9096397/pkg/sql/sem/eval/context.go#L308)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: while reading this i found it super confusing to distinguish this internal `*rand.Rand` from the  one in `connExecutor.rng.internal`. AFAICT, they actually are completely unrelated, right?
</blockquote></details>

Yes, these are totally unrelated. I switched to having internal RNGFactory that should make it less confusing with `connExecutor.rng.internal`.
___
*[`pkg/sql/sem/eval/context.go` line 314 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuVc0h-4ToBJ9l14r5a:-Nu_eDys6P0F1h_uzOwN:b-lqnpbn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c39bacd595b054f529997945ff20349ee9096397/pkg/sql/sem/eval/context.go#L314)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i find this a bit harder to follow now. in this version `RNGFactory` is just a wrapper around `*rand.Rand`, and there's no helpers to initialize the rng. i might be missing something -- what does the RNGFactory buy us now?
</blockquote></details>

Ok, I reverted some of my refactors. The main thing that it was providing is the ability to pass the instantiation of `*rand.Rand` back to the connExecutor - otherwise, we'd have to use `**rand.Rand` I think.
___
*[`pkg/sql/sem/eval/context.go` line 323 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuVUjZ3EwOtCk4kx15E:-Nu_eRzC1ORIwjaQLPS1:b-f709av) ([raw file](https://github.com/cockroachdb/cockroach/blob/c39bacd595b054f529997945ff20349ee9096397/pkg/sql/sem/eval/context.go#L323)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm not sure how much to be worried about this, but if the initialization of the mutex itself is racy, does that introduce the possibility of a deadlock? (e.g. if the reference changes in between the call to lock/unlock)
</blockquote></details>

I removed all new synchronization.
___
*[`pkg/sql/sem/eval/context.go` line 344 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuVUbDB07ggGfRVLVfN:-Nu_eWmO7CkvvKKBJ_fN:bdnb320) ([raw file](https://github.com/cockroachdb/cockroach/blob/c39bacd595b054f529997945ff20349ee9096397/pkg/sql/sem/eval/context.go#L344)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i'm more worried about this change now. it's using a mutex on a potentially hot path of query execution. i'm guessing a benchmark of `SELECT random() FROM generate_series(1, 1000000) ` would perform significantly worse now. a similar realistic example is a query that has `ORDER BY random()`

even if true, i'm not sure what to make of that though... it definitely is the case that `rand.Rand` is not supposed to be used concurrently.

relatedly, using the mutex here doesn't actually make things deterministic, right? (since there's no guarantees about the order it gets accessed by different subqueries.)
</blockquote></details>

It's true that this would have performance impact on execution time of `random()` builtin. I removed all the new synchronization since my goal is avoid redundant allocations.

Using the mutex wouldn't really make things deterministic, it'd only prevent data races when accessing the RNG.

Also, we currently don't use the same seed for `random()` if it is executed on the remote nodes (i.e. the query is distributed). We'd need to mark the builtin as distsql-blocklisted if we wanted to ~do~avoid that.
___
*[`pkg/sql/sem/eval/context.go` line 349 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuVekn05bL2ZlHcYINN:-Nu_eq3kDp_hZo2MAiYI:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c39bacd595b054f529997945ff20349ee9096397/pkg/sql/sem/eval/context.go#L349)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i find this a bit confusing, since sometimes it actually is updating the seed of the `internal` RNG.
</blockquote></details>

Removed.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
@rafiss does this look good to you?

rafiss reviewed (approved):
lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-:-Nwgn5LP6s0JoZ4HcBPh:b-7nbn20)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/sem/eval/context.go` line 344 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121523#-NuVUbDB07ggGfRVLVfN:-Nwgm_LN5qO-c6B5fnrw:b-5rapo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c39bacd595b054f529997945ff20349ee9096397/pkg/sql/sem/eval/context.go#L344)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

It's true that this would have performance impact on execution time of `random()` builtin. I removed all the new synchronization since my goal is avoid redundant allocations.

Using the mutex wouldn't really make things deterministic, it'd only prevent data races when accessing the RNG.

Also, we currently don't use the same seed for `random()` if it is executed on the remote nodes (i.e. the query is distributed). We'd need to mark the builtin as distsql-blocklisted if we wanted to do that.
</blockquote></details>

good point about remote nodes, i hadn't considered that. sounds like it's not worth making it fully deterministic 


<!-- Sent from Reviewable.io -->
